### PR TITLE
[Fix #2175]Adding count process intance and count user task queries

### LINKED ISSUE #2175: Include total number of matching records if the underlying data store suppports that
### Description

Currently, user does not know (cannot calculate) the total number of pages. Although this info is not strictly required to implement a pagination UI (see, for example, Google), it is also true that relational dbs can retrieve that info pretty quickly in some cases. 

### Implementation ideas

The proposal is to add an optional field that might be populated if the total number of records can be calculated by the underlying datastore investing a reasonable performance cost. Note that UIs using Data index cannot rely on this number being present, they should have the "unknown" option for the total number of records.

### PR DESCRIPTION
**Pull request by fjtirado**
Fix https://github.com/apache/incubator-kie-kogito-apps/issues/2175

### CODE REVIEW

File: data-index/data-index-graphql/src/main/resources/basic.schema.graphqls [hunk 1]
```diff
@@ -69,7 +69,6 @@ input ProcessDefinitionArgument {
     serviceUrl: StringArgument
     description: StringArgument
     type: StringArgument
-    metadata: JSON
```
Comment (fjtirado) [inline data-index/data-index-graphql/src/main/resources/basic.schema.graphqls:72]: Here we are preventing json properties to appear int he contextual graphql help in case the datastore does not support it

**Review by gmunozfe — CHANGES_REQUESTED** [review]
Missing `.rat-excludes` entries for schema.graphqls files.

Also 2 failing tests:
```
apache_incubator-kie-kogito-apps/apps-integration-tests/integration-tests-jobs-service/integration-tests-jobs-service-quarkus/integration-tests-jobs-service-quarkus-embedded/src/test/java/org/kie/kogito/it/jobs/SwitchStateTimeoutsIT.java:1 | java.lang.IllegalStateException: Unable to determine the status of the running process. See the above logs for details
apache_incubator-kie-kogito-apps/data-index/data-index-service/data-index-service-postgresql/src/test/java/org/kie/kogito/index/service/PostgreSqlIndexingServiceIT.java:1 | Assertion condition defined as a org.kie.kogito.index.service.AbstractIndexingServiceIT 1 expectation failed. JSON path data.Jobs[0].lastUpdate doesn't match. Expected: is "20[25](https://github.com/apache/incubator-kie-kogito-apps/actions/runs/13504547221/job/37730970537?pr=2196#step:6:27)-02-24T18:42:38.016Z"   Actual: 2025-02-24T18:42:38.017Z  within 30 seconds.
```

File: data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java [hunk 2]
```diff
@@ -195,4 +191,21 @@ private List<Predicate> getRecursivePredicate(AttributeFilter<?> filter, Root<E>
...
     }
 
+    @Override
+    public long count() {
+        CriteriaBuilder builder = repository.getEntityManager().getCriteriaBuilder();
+        CriteriaQuery<Long> criteriaQuery = builder.createQuery(Long.class);
+        Root<E> root = criteriaQuery.from(entityClass);
+        criteriaQuery.select(builder.count(root));
+        addWhere(builder, criteriaQuery, root);
+        return repository.getEntityManager().createQuery(criteriaQuery).getSingleResult();
```
Comment (gmunozfe) [inline data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java:196]: `getSingleResult` could return null. In that case, it would be better to return 0L, wdyt?

File: data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java [hunk 3]
```diff
@@ -195,4 +191,21 @@ private List<Predicate> getRecursivePredicate(AttributeFilter<?> filter, Root<E>
...
+    }
+
+    private <V> void addWhere(CriteriaBuilder builder, CriteriaQuery<V> criteriaQuery, Root<E> root) {
+        if (filters != null && !filters.isEmpty()) {
+            List<Predicate> predicates = getPredicates(builder, root);
+            criteriaQuery.where(predicates.toArray(new Predicate[] {}));
```
Comment (gmunozfe) [inline data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java]: you can also use `criteriaQuery.where(predicates.toArray(Predicate[]::new));`
but it's up to you, no big difference at all

File: data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java [hunk 4]
```diff
@@ -205,10 +228,22 @@ protected <K, T> List<T> executeAdvancedQueryForCache(StorageFetcher<K, T> cache
...
-
         return query.execute();
     }
 
+    protected <K, T> long executeCount(StorageFetcher<K, T> cache, DataFetchingEnvironment env) {
+        return setupQuery(cache, env).count();
+    }
+
+    private <K, T> Query<T> setupQuery(StorageFetcher<K, T> cache, DataFetchingEnvironment env) {
+        Objects.requireNonNull(cache, "Cache not found");
```
Comment (gmunozfe) [inline data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java]: Shouldn't we require that `env` is not null as well?
`Objects.requireNonNull(env, "DataFetchingEnvironment cannot be null");`

File: data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java [hunk 5]
```diff
@@ -205,10 +228,22 @@ protected <K, T> List<T> executeAdvancedQueryForCache(StorageFetcher<K, T> cache
...
+        String inputTypeName = ((GraphQLNamedType) env.getFieldDefinition().getArgument("where").getType()).getName();
+        Query<T> query = cache.query();
+        Map<String, Object> where = env.getArgument("where");
+        query.filter(GraphQLQueryParserRegistry.get().getParser(inputTypeName).apply(where));
```
Comment (gmunozfe) [inline data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java]: I'm wondering if we could make any performance optimization for same cache and env, but probably it's out of the scope of this feature and can be handle later.

**Review by gmunozfe — APPROVED** [review]
Looks good to me, but needs some integration tests to validate the count for process instance and user task and also test some edge cases.
Rest of comments are minor, just check the possible null return for getSingleResult.
Nice work @fjtirado !

File: data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java (hunk 3 above)
Comment (fjtirado) [reply data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java]: I do not change the original code, but actually, it should be predicates.toArray(new Predicate[predicates.size()])

File: data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java (hunk 4 above)
Comment (fjtirado) [reply data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java]: I did not change the original code, just moved it (so it appears in the review)  Env is never going to be null, so there is not need   to check it (I cannot think of an scenario where cache is null, but since the code was already there and was unrelated with the PR, I prefer to not change that, just in case)

File: data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java (hunk 5 above)
Comment (fjtirado) [reply data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java]: I think the query object cannot be reused between different request.

File: data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java (hunk 4 above)
Comment (fjtirado) [reply data-index/data-index-graphql/src/main/java/org/kie/kogito/index/graphql/AbstractGraphQLSchemaManager.java]: After taking another view, I decided to change the original code to not check for nulls that are not going to be null outside test contest and check for nulls that might actually be there (although they apparently not in current scenarios)

File: data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java (hunk 3 above)
Comment (fjtirado) [reply data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java]: In fact, there it not need to collect items to List, they can be collected to array directly

File: data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java (hunk 2 above)
Comment (fjtirado) [reply data-index/data-index-storage/data-index-storage-jpa-common/src/main/java/org/kie/kogito/index/jpa/storage/JPAQuery.java:196]: Added a new test that proof that getSingleResult return 0 if there are not items, so nothing to change here

**Review by wmedvede — APPROVED** [review]

**Comment by kie-ci3** [comment]
**PR job** `#15` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-apps -u https://github.com/apache/incubator-kie-kogito-apps/pull/2196 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2196/15/display/redirect

**Test results:**
- PASSED: 1826
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-apps-pr/job/PR-2196/15/testReport/org.kie.kogito.jobs.embedded/EmbeddedJobsServiceTest/testJobService/">org.kie.kogito.jobs.embedded.EmbeddedJobsServiceTest.testJobService</a></summary>
java.util.ConcurrentModificationException<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1513)<br/>	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)<br/>	at org.kie.kogito.jobs.embedded.EmbeddedJobsServiceTest.testJobService(EmbeddedJobsServiceTest.java:99)<br/>	at java.base/java.lang.reflect.Method.invoke(Method.java:568)<br/>	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:973)<br/>	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:823)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)<br/>
</details>

**Comment by pefernan** [comment]
@fjtirado nice addition! Would it be possible to have it for jobs too?

**Comment by fjtirado** [comment]
> @fjtirado nice addition! Would it be possible to have it for jobs too?

Sure, I have opened https://github.com/apache/incubator-kie-kogito-apps/issues/2197 to do that.
